chia/cmds/cmd_classes.py

```diff
@@ -0,0 +1,322 @@
+
+
+@dataclass(frozen=True)
+class _CommandParsingStage:
+    my_dataclass: Type[ChiaCommand]
+    my_option_decorators: List[Callable[[SyncCmd], SyncCmd]]
+    my_subclasses: Dict[str, _CommandParsingStage]
+    my_kwarg_names: List[str]
+    _needs_context: bool
+
+    def needs_context(self) -> bool:
+        if self._needs_context:
+            return True
+        else:
+            return any([subclass.needs_context() for subclass in self.my_subclasses.values()])
+
+    def get_all_option_decorators(self) -> List[Callable[[SyncCmd], SyncCmd]]:
+        all_option_decorators: List[Callable[[SyncCmd], SyncCmd]] = self.my_option_decorators
+        for subclass in self.my_subclasses.values():
+            all_option_decorators.extend(subclass.get_all_option_decorators())
+        return all_option_decorators
+
+    def initialize_instance(self, **kwargs: Any) -> ChiaCommand:
+        kwargs_to_pass: Dict[str, Any] = {}
+        for kwarg_name in self.my_kwarg_names:
+            kwargs_to_pass[kwarg_name] = kwargs[kwarg_name]
+
+        for subclass_arg_name, subclass in self.my_subclasses.items():
+            kwargs_to_pass[subclass_arg_name] = subclass.initialize_instance(**kwargs)
+
+        return self.my_dataclass(**kwargs_to_pass)
+
+    def apply_decorators(self, cmd: SyncCmd) -> SyncCmd:
+        cmd_to_return = cmd
+        if self.needs_context():
+
+            def strip_click_context(func: SyncCmd) -> SyncCmd:
+                def _inner(ctx: click.Context, **kwargs: Any) -> None:
+                    context: Dict[str, Any] = ctx.obj if ctx.obj is not None else {}
+                    func(context=context, **kwargs)
+
+                return _inner
+
+            cmd_to_return = click.pass_context(strip_click_context(cmd_to_return))
+
+        for decorator in self.get_all_option_decorators():
+            cmd_to_return = decorator(cmd_to_return)
+
+        return cmd_to_return
+
+
+def _generate_command_parser(cls: Type[ChiaCommand]) -> _CommandParsingStage:
+    option_decorators: List[Callable[[SyncCmd], SyncCmd]] = []
+    kwarg_names: List[str] = []
+    subclasses: Dict[str, _CommandParsingStage] = {}
+    needs_context: bool = False
+
+    hints = get_type_hints(cls)
+    _fields = fields(cls)  # type: ignore[arg-type]
+
+    for _field in _fields:
+        field_name = _field.name
+        if (
+            isinstance(hints[field_name], type)
+            and hasattr(hints[field_name], "__command_helper__")
+            and hints[field_name].__command_helper__
+        ):
+            subclasses[field_name] = _generate_command_parser(hints[field_name])
+        else:
+            if field_name == "context":
+                if hints[field_name] != Context:
+                    raise ValueError("only Context can be the hint for variables named 'context'")
+                else:
+                    needs_context = True
+                    kwarg_names.append(field_name)
+                    continue
```

Comment:
I haven't looked over the rest below for this opportunity, but I feel like there may be good chances to de-nest this a bit more similar to this.

```
        elif field_name == "context":
            if hints[field_name] != Context:
                raise ValueError("only Context can be the hint for variables named 'context'")
            needs_context = True
            kwarg_names.append(field_name)
        else:
```